### PR TITLE
feat: BA gambles notifier

### LINKED ISSUE #145: New notifier: Barbarian Assault high level gambles
Hi from Casual BA! I've noticed that this plugin supports a bunch of neat webhooks and was interested in the possibility of automatically sending players' high gamble counts to our Discord, since we have a channel for players to submit these manually. Gambles are a big deal in our community as we use it to determine ranks, maintain our own unofficial hiscores, and award discord badges for certain milestones. If this feature is deemed appropriate for this plugin, I'd be down to implement it.

### PR DESCRIPTION
**Pull request by equirs**
fixes #145 

Requires the next core release in order to use the new `MESBOX` chat message type

### CODE REVIEW

File: src/test/java/dinkplugin/notifiers/LevelNotifierTest.java [hunk 1]
```diff
@@ -97,7 +97,7 @@ void testNotifyTwo() {
             PRIMARY_WEBHOOK_URL,
             false,
             NotificationBody.builder()
-                .text(PLAYER_NAME + " has levelled Agility to 5 and Attack to 99")
+                .text(PLAYER_NAME + " has levelled Agility to 5 and Hunter to 99")
```
Comment (equirs) [inline src/test/java/dinkplugin/notifiers/LevelNotifierTest.java:100]: I noticed that these tests don't actually verify the text on `NotificationBody` since it's a transient field. When running the tests with non-transient text, I found that this test was failing. I won't mess with `NotificationBody` since I guess there's a good reason why it uses transient

File: src/main/java/dinkplugin/DinkPluginConfig.java [hunk 2]
```diff
@@ -1004,4 +1023,70 @@ default String diaryNotifyMessage() {
...
+    @ConfigItem(
+        keyName = "gambleInterval",
+        name = "Notify Interval",
+        description = "Interval between when a notification should be sent",
+        position = 122,
+        section = gambleSection
+    )
+    @Range(min = 1)
+    default int gambleInterval() {
+        return 100;
```
Comment (equirs) [inline src/main/java/dinkplugin/DinkPluginConfig.java:1057]: I chose a higher default here compared to other intervals after bringing this up with the Casual BA admins. There was some concern that people would keep it at 1, even if we instructed them to set it to 100 or more (our rank-ups happen at 100 / 500 / 1000 gambles, so this interval seemed to be the most reasonable).

File: build.gradle.kts [hunk 3]
```diff
@@ -39,7 +39,7 @@ dependencies {
 }
 
 group = "dinkplugin"
-version = "1.2.2"
+version = "1.2.3"
```
Comment (Felanbird) [inline build.gradle.kts:42]: this will be done in a separate PR

File: build.gradle.kts [hunk 4]
```diff
@@ -21,7 +21,7 @@ dependencies {
     // this version of annotations is verified by runelite
     compileOnly(group = "org.jetbrains", name = "annotations", version = "23.0.0")
 
-    val runeLiteVersion = "latest.release"
+    val runeLiteVersion = "1.9.9-SNAPSHOT" // todo revert
```
Comment (Felanbird) [inline build.gradle.kts:24]: this todo so its not missed

File: src/test/java/dinkplugin/notifiers/LevelNotifierTest.java (hunk 1 above)
Comment (iProdigy) [reply src/test/java/dinkplugin/notifiers/LevelNotifierTest.java:100]: Good catch, `transient` is important but we should add `@EqualsAndHashCode.Include` on the field

**Review by Felanbird — CHANGES_REQUESTED** [review]
this at the top of `CHANGELOG.md`
`- Major: Add Barbarian Assault Gambling notifications. (#150)`

**Comment by Felanbird** [comment]
@equirs from the current state of this PR - do you think expanding to support `low` & `medium` gambles would be simple?

It doesn't have to be done but it would better complete the feature if done. It could have a `Min Tier` similar to clues/diaries/combat tasks that's set to `High` by default

**Comment by equirs** [comment]
> @equirs from the current state of this PR - do you think expanding to support `low` & `medium` gambles would be simple?
> 
> It doesn't have to be done but it would better complete the feature if done. It could have a `Min Tier` similar to clues/diaries/combat tasks that's set to `High` by default

Implementing would be straightforward, but I doubt anyone would use it since there are no notable rewards, no pet rolls, etc. There's not even a chat command for it in RuneLite (`!gc` returns high gambles) nor is there a request for such a feature. But I believe they are tracked, just have to test it out when I get a chance

**Comment by Felanbird** [comment]
> RuneLite (`!gc` returns high gambles) nor is there a request for such a feature. 

That's fine then, if it's straightforward then it can be implemented in the future if needed 👍

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        String itemName = rawItem;
+        int itemQuantity = 1;
+        Matcher quantityMatcher = ITEM_QUANTITY_REGEX.matcher(rawItem);
+        if (quantityMatcher.matches()) {
+            itemName = quantityMatcher.group(1).trim();
+            itemQuantity = Integer.parseInt(quantityMatcher.group(2));
+        }
+        String tertiary = matcher.group(2);
+        if (tertiary != null) {
+            tertiary = tertiary.replaceAll("!", "").trim();
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:93]:
```suggestion
            tertiary = tertiary.replace("!", "").trim();
```
avoids more expensive regex method... alternatively you can use `StringUtils.removeEnd(tertiary, "!")`

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+@Slf4j
+@Singleton
+public class GambleNotifier extends BaseNotifier {
+    // itemName (x quantity)![ tertiaryReward!] gc: n
+    private static final Pattern GAMBLE_REGEX = Pattern.compile(
+        "^(.+?)!\\s*(.+!)?\\s+High level gamble count: (\\d+).*"
+    );
+    private static final Pattern ITEM_QUANTITY_REGEX = Pattern.compile("^(.+?)\\(x (\\d+)\\)$");
+    private static final List<String> RARE_LOOT = ImmutableList.of("dragon chainbody", "dragon med helm");
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:36]: Let's make this a `Collection<String>` since it's really describing a set... but I'm fine with keeping the concrete implementation as list given the likely faster contains check in practice due to low n (despite theoretically worse algo complexity)

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    @Value
+    @VisibleForTesting
+    static class ParsedData {
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:40]: Can we move this class definition to the end of the file

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        if (!isEnabled()) return;
+        ParsedData data = parse(message);
+        if (data != null) {
+            handleNotify(data);
+        }
+    }
+
+    @Nullable
+    @VisibleForTesting
+    static ParsedData parse(String message) {
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:80]: this can also be moved down as the last method before the inner class definition

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        }
+        String player = Utils.getPlayerName(client);
+        String message = StringUtils.replaceEach(
+            messageFormat,
+            new String[] { "%USERNAME%", "%COUNT%", "%LOOT%" },
+            new String[] { player, String.valueOf(data.gambleCount), lootSummary(data) }
+        );
+        createMessage(config.gambleSendImage(), NotificationBody.builder()
+            .text(message)
+            .extra(new GambleNotificationData(data.gambleCount, serializeItems(data)))
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:116]: Let's assign `serializeItems(data)` to a local variable so `SerializedItemStack`s can be passed to `#lootSummary` (instead of `ParsedData`)

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            .type(NotificationType.BARBARIAN_ASSAULT_GAMBLE)
+            .build());
+    }
+
+    // ex. "watermelon seed x 50, clue scroll (elite)"
+    private String lootSummary(ParsedData data) {
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:122]: With this converted to accept `SerializedItemStack`s, we can stringify items with the same format as clue/loot notifier: https://github.com/pajlads/DinkPlugin/blob/master/src/main/java/dinkplugin/notifiers/ClueNotifier.java#L136 (with new-line as delimeter)

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        return data.itemsWithQuantities()
+            .map(p -> {
+                Integer count = p.getRight();
+                return p.getLeft().toLowerCase() + (count > 1 ? " x " + count : "");
+            })
+            .collect(Collectors.joining(", "));
+    }
+
+    private List<SerializedItemStack> serializeItems(ParsedData data) {
+        return data.itemsWithQuantities()
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:132]: With the proposed changes, `itemsWithQuantities` would only be called once, so we can consider removing the method altogether

File: src/main/java/dinkplugin/notifiers/data/GambleNotificationData.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+
+import java.util.List;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class GambleNotificationData extends NotificationData {
+    int gambleCount;
+    List<SerializedItemStack> loot;
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/data/GambleNotificationData.java:12]:
```suggestion
    List<SerializedItemStack> items;
```

for consistency with clue/loot notifiers

**Review by iProdigy — COMMENTED** [review]
Great work on this! Thanks again for taking the time to contribute to Dink & good job on the unit tests

Have a handful of nitpicks below, but overall happy to see this addition (and personally don't require the gamble level to be done in this PR)

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java (hunk 10 above)
Comment (equirs) [reply src/main/java/dinkplugin/notifiers/GambleNotifier.java:122]: Was looking for a util function for this. Maybe it's worth adding it to `ItemUtils`?

Comment (iProdigy) [reply src/main/java/dinkplugin/notifiers/GambleNotifier.java:122]: good idea 👍

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java (hunk 6 above)
Comment (equirs) [reply src/main/java/dinkplugin/notifiers/GambleNotifier.java:36]: Now that you bring it up, it doesn't really feel right as a list

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Slf4j
+@Singleton
+public class GambleNotifier extends BaseNotifier {
+    // itemName (x quantity)![ tertiaryReward!] gc: n
+    private static final Pattern GAMBLE_REGEX = Pattern.compile("^(.+?)!\\s*(.+!)?\\s+High level gamble count: (\\d+).*");
+    private static final Pattern ITEM_QUANTITY_REGEX = Pattern.compile("^(.+?)\\(x (\\d+)\\)$");
+    private static final Collection<String> RARE_LOOT = ImmutableSet.of("dragon chainbody", "dragon med helm");
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:35]: Is [pet penance queen](https://oldschool.runescape.wiki/w/Barbarian_Assault/Rewards#Tertiary_High_Gamble_Rewards) covered by this approach?

Comment (equirs) [reply src/main/java/dinkplugin/notifiers/GambleNotifier.java:35]: Unfortunately, the pet doesn't show up anywhere in the message upon receiving it. I learned this yesterday from experience :frowning_face:

Comment (Felanbird) [reply src/main/java/dinkplugin/notifiers/GambleNotifier.java:35]: Did you receive the normal `You have a funny feeling...` chat message? If so then it will just be handled by our existing pet notifier.

Comment (equirs) [reply src/main/java/dinkplugin/notifiers/GambleNotifier.java:35]: Yeah, the normal pet drop game message appears, so it will be handled by the pet notifier

Comment (iProdigy) [reply src/main/java/dinkplugin/notifiers/GambleNotifier.java:35]: > Unfortunately, the pet doesn't show up anywhere in the message upon receiving it. I learned this yesterday from experience ☹️

Ah that's unfortunate - perhaps we can add it to the rare set anyways (in case jagex changes this) with a comment explaining it currently doesn't occur

**Review by pajlada — CHANGES_REQUESTED** [review]
Small nitpicks, overall looks good!

File: src/main/java/dinkplugin/DinkPluginConfig.java [hunk 14]
```diff
@@ -1004,4 +1023,70 @@ default String diaryNotifyMessage() {
         return "%USERNAME% has completed the %DIFFICULTY% %AREA% Achievement Diary, for a total of %TOTAL% diaries completed";
     }
 
+    @ConfigItem(
+        keyName = "gambleEnabled",
+        name = "Enable BA Gamble",
+        description = "Enable notifications for Barbarian Assault gambles",
```
Comment (iProdigy) [inline src/main/java/dinkplugin/DinkPluginConfig.java:1029]:
```suggestion
        description = "Enable notifications for Barbarian Assault high-level gambles",
```

File: src/main/java/dinkplugin/notifiers/GambleNotifier.java [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+            new String[] { player, String.valueOf(data.gambleCount), lootSummary(items) }
+        );
+        createMessage(config.gambleSendImage(), NotificationBody.builder()
+            .text(message)
+            .extra(new GambleNotificationData(data.gambleCount, items))
+            .type(NotificationType.BARBARIAN_ASSAULT_GAMBLE)
+            .build());
+    }
+
+    private String lootSummary(List<SerializedItemStack> items) {
```
Comment (iProdigy) [inline src/main/java/dinkplugin/notifiers/GambleNotifier.java:85]:
```suggestion
    private static String lootSummary(List<SerializedItemStack> items) {
```

**Review by iProdigy — COMMENTED** [review]
my last nits 😄

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,6 @@
 ## Unreleased
 
+- Major: Add Barbarian Assault Gambling notifications. (#150)
```
Comment (Felanbird) [inline CHANGELOG.md:3]:
```suggestion
- Major: Add Barbarian Assault high-level Gambling notifications. (#150)
```

Throw this nitpick in since I wrote this Changelog while thinking about the possible low&medium implementation

**Comment by iProdigy** [comment]
Appearance of rich embeds (from tests)

image

**Comment by pajlada** [comment]
The icon in the top-right is a lot larger than our other notifiers, is this a concern we care about? personally I think this looks fine

**Comment by equirs** [comment]
> The icon in the top-right is a lot larger than our other notifiers, is this a concern we care about? personally I think this looks fine

open to using another icon, I just couldn't find one similar enough that symbolizes BA.

I tested the embed as well a few days ago on some gambles using an interval of 10:

**Comment by iProdigy** [comment]
I also couldn't find a smaller icon & the image is non-free so I don't think we have the license to modify it (so I'm fine with this)

**Review by iProdigy — APPROVED** [review]

**Review by pajlada — CHANGES_REQUESTED** [review]
Looks good to me! Now we wait for the 1.9.9 release for MESBOX - marking as "Request changes" even though this is technically an "Approve" just so it doesn't get accidentally merged

**Review by Felanbird — APPROVED** [review]

File: src/main/java/dinkplugin/DinkPlugin.java [hunk 17]
```diff
@@ -153,6 +155,10 @@ public void onChatMessage(ChatMessage message) {
                 // do nothing
                 break;
         }
+
+        if (msgType == ChatMessageType.MESBOX) {
```
Comment (iProdigy) [inline src/main/java/dinkplugin/DinkPlugin.java:159]: Could you move this into the new switch-case statement above

**Review by pajlada — APPROVED** [review]

**Review by iProdigy — APPROVED** [review]
Thank you again for the contribution & bearing with our nits!
